addons/mail/models/discuss/ir_binary.py

```diff
@@ -21,4 +21,27 @@ def _find_record_check_access(self, record, access_token, field):
                     domain.append(("channel_member_ids", "any", [("guest_id", "=", current_guest.id)]))
                 if self.env["discuss.channel"].search_count(domain, limit=1):
                     return record.sudo()
+                if current_guest:
+                    # access to the record is allowed if there are common channels
+                    # between record as a guest or partner and current guest
+                    if record._name == "res.partner":
+                        domain = [("author_id", "=", record.id), ("model", "=", "discuss.channel")]
+                    else:
+                        domain = [
+                            ("author_guest_id", "=", record.id),
+                            ("model", "=", "discuss.channel"),
+                        ]
+                    messages = self.env["mail.message"].search_fetch(domain, ["res_id"])
+                    if messages and self.env["discuss.channel"].search_count(
+                        [
+                            ("id", "in", messages.mapped("res_id")),
+                            (
+                                "channel_member_ids",
+                                "any",
+                                [("guest_id", "=", current_guest.id)],
+                            ),
+                        ],
+                        limit=1,
+                    ):
```

Comment:
Normally just access of message is enough, all of this can be removed.
If we want to be extra safe (or future proof compatible with portal) we can call `_get_with_access` on the messages and see if any is allowed.